js_modules/dagit/packages/ui/src/components/MiddleTruncate.tsx

```diff
@@ -9,14 +8,87 @@ interface Props {
+  // taken place yet.
   const [targetStyle, setTargetStyle] = React.useState<TargetStyle | null>(null);
-  const ref = React.useRef<HTMLDivElement>(null);
 
-  const truncated = useFixedSpan(text, targetStyle);
+  // Determine the width available to the text container. This renders a long string into
+  // the container to allow quickly measuring the maximum width available, which will then
+  // be used to determine the truncated string.
+  const [determineAvailableWidth, setDetermineAvailableWidth] = React.useState(true);
+
+  const container = React.useRef<HTMLDivElement>(null);
+
+  // Track whether we have just evaluated the avilable width of the container. If so,
+  // we want to ignore any resize observations temporarily, to avoid an observer->resize loop.
+  const justDetermined = React.useRef(false);
+
+  // Given the target font style and allotted width, calculate the largest possible middle-
+  // truncated string. This hook returns null if there is no style available yet.
+  const truncated = useMiddleTruncatedText(text, targetStyle);
+
+  // When we are rendering the "fill" string to determine the available width, use a layout
+  // effect to perform the measurement. Get the computed font style as well -- both of these
+  // will be used to draw the text string onto a canvas context to find the appropriate
+  // truncated string.
+  React.useLayoutEffect(() => {
+    if (container.current && determineAvailableWidth) {
+      justDetermined.current = true;
+      setTargetStyle({
+        width: container.current.getBoundingClientRect().width,
+        font: getFont(container.current),
+      });
+
+      // Done with the fill string. On the next render, we will use the truncated string
+      // as determined by `useMiddleTruncatedText`.
+      setDetermineAvailableWidth(false);
+    }
+  }, [determineAvailableWidth]);
+
+  // As noted above, we are using a ref to track whether we have just finished determining the
+  // maximum available width of the container. If so, clear the value after a brief delay. This
+  // prevents our resize observer from responding to the resize that will occur when we re-render
+  // the text in the DOM, which could otherwise lead to an observer->resize loop.
+  const shouldClear = justDetermined.current;
+  React.useEffect(() => {
+    let timer: ReturnType<typeof setTimeout> | undefined;
+    if (shouldClear) {
+      timer = setTimeout(() => {
+        justDetermined.current = false;
+      }, 20);
+    }
+    return () => {
+      timer && clearTimeout(timer);
+    };
+  }, [shouldClear]);
+
+  // If the container has just been resized -- but NOT due to a width difference between rendering "fill"
+  // text and our actual text -- then re-render the "fill" in order to perform a new measurement and
+  // recalculate the middle truncation.
+  const observerFn = React.useCallback(() => {
+    if (!justDetermined.current) {
+      setDetermineAvailableWidth(true);
+    }
+  }, []);
+
```

Comment:
This looks good to me! If the flag + setTimeout isn't sufficient and we keep seeing problems with ResizeObserver loops, I think we could also fix this by detatching + re-attaching the observer on each render (so  the render itself can't trigger it?) 

edit: It's actually kinda strange this is happening - I wonder if removing the React.useCallback would actually fix this, because it'd force useResizeObserver to detach + reattach on each render... the useResizeObserver hook might also need to mount in a useLayoutEffect instead of a useEffect :-/